src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsSerializer.cs

```diff
@@ -0,0 +1,262 @@
+            }
+            else if (option == CompletionOptions.TriggerOnTyping)
+            {
+                languagePreference.fAutoListMembers = Convert.ToUInt32((bool)value ? 1 : 0);
+            }
+            else if (option == SignatureHelpOptions.ShowSignatureHelp)
+            {
+                languagePreference.fAutoListParams = Convert.ToUInt32((bool)value ? 1 : 0);
+            }
+            else if (option == NavigationBarOptions.ShowNavigationBar)
+            {
+                languagePreference.fDropdownBar = Convert.ToUInt32((bool)value ? 1 : 0);
+            }
+            else if (option == BraceCompletionOptions.EnableBraceCompletion)
+            {
+                languagePreference.fBraceCompletion = Convert.ToUInt32((bool)value ? 1 : 0);
+            }
+            else
+            {
+                throw new ArgumentException("Unexpected option.", nameof(option));
+            }
+        }
+
+        public bool TryFetch(OptionKey optionKey, out object value)
+        {
+            // This particular serializer is a bit strange, since we have to initially read things out on the UI thread.
+            // Therefore, we refresh the values in the constructor, meaning that this should never get called for our values.
+
+            Contract.ThrowIfTrue(_supportedOptions.Contains(optionKey.Option));
+
+            value = null;
+            return false;
+        }
+
+        public bool TryPersist(OptionKey optionKey, object value)
+        {
+            if (!_supportedOptions.Contains(optionKey.Option))
+            {
+                value = null;
+                return false;
+            }
+
+            Guid languageServiceGuid;
+
+            if (!_languageMap.TryGetValue(optionKey.Language, out languageServiceGuid))
+            {
+                value = null;
+                return false;
+            }
+
+            var languagePreferences = new LANGPREFERENCES3[1];
+            languagePreferences[0].guidLang = languageServiceGuid;
+            Marshal.ThrowExceptionForHR(_textManager.GetUserPreferences4(null, languagePreferences, null));
+
+            SetValueForOption(optionKey.Option, ref languagePreferences[0], value);
+            SetUserPreferencesMaybeAsync(languagePreferences);
+
+            // Even if we didn't call back, say we completed the persist
+            return true;
+        }
+
+        private void SetUserPreferencesMaybeAsync(LANGPREFERENCES3[] languagePreferences)
+        {
+            if (IsForeground())
+            {
+                Marshal.ThrowExceptionForHR(_textManager.SetUserPreferences4(pViewPrefs: null, pLangPrefs: languagePreferences, pColorPrefs: null));
+            }
+            else
+            {
+                Task.Factory.StartNew(() => this.SetUserPreferencesMaybeAsync(languagePreferences), CancellationToken.None, TaskCreationOptions.None, ForegroundThreadAffinitizedObject.CurrentForegroundThreadData.TaskScheduler);
```

Comment:
does this guarantee ordering? if 2 feature change options from background, will option be saved FIFO manner?